subsys/shell/shell_rtt.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2018 Makaio GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <shell/shell_rtt.h>
+#include <init.h>
+#include <rtt/SEGGER_RTT.h>
+
+SHELL_RTT_DEFINE(shell_transport_rtt);
+SHELL_DEFINE(rtt_shell, "rtt:~$ ", &shell_transport_rtt, 10,
+	     SHELL_FLAG_OLF_CRLF);
+
+
+static struct k_thread rtt_rx_thread;
+static K_THREAD_STACK_DEFINE(rtt_rx_stack, 1024);
+
+static void shell_rtt_rx_process(struct shell_rtt *sh_rtt)
+{
+	unsigned int count;
+	while (true)
+	{
+		SEGGER_RTT_LOCK();
+		count = SEGGER_RTT_ReadNoLock(0, sh_rtt->rx, sizeof(sh_rtt->rx));
+
+		if(count > 0)
+		{
+			sh_rtt->rx_cnt = 1;
```

Comment:
I would propose to increase `u8_t rx[1];` to 5 bytes and copy more bytes from `SEGGER_RTT_ReadNoLock`. 
You will be able to handle escape code sequence within one shell function call. For example if someone will press an `arrow`, or `backspace` or `tab` shell reaction will be much faster.

Reply:
Actually, I can't test that as it seems when running via nc I always have to hit enter to send data, but I'll change it

Reply:
should rx_cnt == count ?

Reply:
Yes you need to change it:
~~`sh_rtt->rx_cnt = 1;`~~
`sh_rtt->rx_cnt = count;`